text/3371-cargo-target-base-dir.md

```diff
@@ -0,0 +1,402 @@
+
+In the same vein, `cargo metadata` fills the target directory information with the absolute path and make no mention of the template in `CARGO_TARGET_DIR` since it can only be used with a single workspace at once.
+
+### `cargo clean`
+
+Currently, if `CARGO_TARGET_DIR` is set to anything but `target` for a project, `cargo clean` does not delete the `target/` directory if it exists, instead deleting the directory pointed by `CARGO_TARGET_DIR`. The same behavior is used for the templated version: if it set, `cargo clean` deletes `/path/to/<manifest-path-hash>/` and not `target/`.
+
+### Providing forward links
+
+[`targo`][tg] provides forward link (it links from `<workspace>/target` to its own target directory) as a way for existing tools to continue working despite there being no explicit `CARGO_TARGET_DIR` set for them to find the real target directory.
+
+`cargo` does not provide them for regular (untemplated) `CARGO_TARGET_DIR`. This is not a limitation when using the environment variable set globally, since all processes can read it, but it is one when this config is only set on specific calls or via `target-dir` in the config, meaning others tools cannot easily pick it up (and most external tools don't use `cargo-metadata`, which makes them all broken by default, but fixing this situation is not this RFC's purpose).
+
+When a templated `CARGO_TARGET_DIR` is used (in any form) (and not overriden, for example a templated env var overriden by a precise `--target-dir` option), it *will* use a forward link by adding a `target` symlink to the real target directory. This `target` symlink will be in the exact place the real target directory would have been if the templated `CARGO_TARGET_DIR` wasn't set at all.
+
+This has a two big advantages: not breaking external tools and giving easy access to artifacts produced by `cargo build/test/doc` to users (they're in the habit of typing `./target/debug/my-bin`, this would continue working with forward links).
+
+A config option (CLI, `config.toml` and env var), `link-target-dir`, will be introduced to deactivate this behaviour, but it will `true` by default, for the reasons provided in favor of forward links just above.
+
+#### Detailed working of forward links
+
+When creating a forward link `cargo` will first attempt to create a symbolic link (regardless of the platform). If that fails, it will attempt zero or more platform-specific solutions, like junction points on NTFS. If that fails too, a warning will be emitted but this will not prevent the rest of the action to go on: regular calls like `cargo check/clippy/build/test` likely won't need this forward link and after the user has been warned they could either resolve the problem themselves or ignore it, depending on their own use case and domain-specific knowledge.
```

Comment:
> a warning will be emitted but this will not prevent the rest of the action to go on

This could be very noisy for Windows users. I guess its reduced if we tell users that they can turn off forward links but I think I'd rather we fail silently.  Otherwise, people are likely to put this in their project or user config.  In the project config, it  now applies to everyone, regardless of platforms.  In the user config, this makes it harder to share a config across platforms.

Reply:
We could also make it only happen when the actual target directory is created and not after that, that way it will warn once then won't as long as there is no "cargo clean"

Maybe that's still too verbose and we would only warn in `--verbose` mode ?

Reply:
> We could also make it only happen when the actual target directory is created and not after that, that way it will warn once then won't as long as there is no "cargo clean"

That might be sufficient.    We should probably put this down as an Unresolved Question for being determined during stabilization
- warning vs note
- always, on target creation, or with `--verbose`

Reply:
When to create/modify it is also more complicated, as you noted in https://github.com/rust-lang/rfcs/pull/3371#discussion_r1493519124, I'll move that to everything related to `Unresolved Questions` with the difficulties explained

Reply:
I feel like the "only create forward link on target dir creation is an important note to call out (with the reason)

Reply:
I expand on forward links in 6540d11b8e2a2f4def265989a925932843ced8f3, it should hopefully be clearer